lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1586,13 +1616,99 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
       if (!ty->hasReferenceSemantics())
         return false;
 
+      if (auto closureExpr = dyn_cast<ClosureExpr>(inClosure)) {
+        if (auto selfDecl = closureExpr->getCapturedSelfDecl()) {
+          // If this capture is using the name `self` actually referring
+          // to some other variable (e.g. with `[self = "hello"]`)
+          // then implicit self is not allowed.
+          if (!selfDecl->isSelfParamCapture()) {
+            return true;
+          }
+        }
+      }
+
+      if (auto var = dyn_cast<VarDecl>(DRE->getDecl())) {
+        if (!isEnclosingSelfReference(var, inClosure)) {
+          return false;
+        }
+      }
+
       return true;
     }
 
+    static bool
+    implicitWeakSelfReferenceIsValid(DeclRefExpr *DRE,
+                                     const AbstractClosureExpr *inClosure) {
+      ASTContext &Ctx = DRE->getDecl()->getASTContext();
+      auto selfDecl = DRE->getDecl();
+
+      if (!Ctx.LangOpts.isSwiftVersionAtLeast(6)) {
+        // In Swift 5 language modes, the decl of this implicit `DeclRefExpr`
+        // incorrectly refers to the `self` `ParamDecl` of the closure
+        // (which is always non-optional), instead of the actual decl that
+        // an explicit `self` call would refer to. As a workaround, we can
+        // manually lookup what self "should" refer to, and then validate
+        // that decl instead.
+        auto lookupResult = ASTScope::lookupSingleLocalDecl(
+            inClosure->getParentSourceFile(), DeclName(Ctx.Id_self),
+            DRE->getLoc());
```

Comment:
Could you please explain to me again why is this a good idea? The AST would still have the original declaration as a base which means that even though we don't warn the behavior is not what users would expect it to be?

Reply:
In Swift 5 mode, the AST node for an implicit self decl always points to the `ParamDecl` from the outer scope. 

To implement the requirements described in SE-0365, we have to check where `self` was actually defined by the user. There are many different ways to define a `self` property, but the only declarations that permit implicit self are optional unbinding conditions like `guard let self = self` or `if let self = self`. If the user wrote something else (like `guard let self = NewInstanceOfSelf()` as an example), we don't permit implicit self. 

This requires knowing what `self` actually refers to at the given DRE location. Since the `ParamDecl` in the DRE is incorrect in Swift 5 mode, we have to do a separate lookup here. This is not necessary in Swift 6 mode, because the DRE's decl correctly points to the `let self = self` definition earlier on the scope.

Reply:
> even though we don't warn the behavior is not what users would expect it to be

I think the user-facing behavior is still correct here, even though the AST itself is a bit unusual. For escaping closures there are no user-facing behavior changes between between Swift 5 and Swift 6 (other than slightly improved error diagnostics in Swift 6 mode).

Reply:
I'm asking explicitly why suppress the warning in Swift 5 if the `self` is indeed incorrect? That would only affect warnings-as-errors mode otherwise the code would still compile.

Reply:
The warning this suppresses happens in code that is actually well-formed and correct, according to the user-facing language model.

For example:

```swift
buttonTapped { [weak self] in
  guard let self else { return } // WARNING: Property `self` is written to but never read
  dismiss() // implicitly refers to `self.dismiss()`
}
```

Since the implicit self DRE on the `dismiss()` call refers to the `ParamDecl` and not the `guard let self` decl, `VarDeclUsageChecker` will consider the `guard let self` condition unused and will emit an error. This warning is incorrect according to the user-facing language model, and the code will no longer compile if the statement is removed, so the warning need to be suppressed.

Reply:
The warning is technically correct right?  if `self` as declared in the body is indeed not used why would it cause an error?

Reply:
I don't think we can ship SE-0365 with that warning, considering it would be emitted in every case where you use the functionality described in SE-0365.

> The warning is technically correct right? 

I don't think the user needs to know about the details of the underlying AST. I think this warning is incorrect given the user-facing model of how this is intended to work -- that implicit self is permitted in the scope once `self` has been unwrapped using a conditional statement.

Reply:
I think other option would be to delay SE-0365 entirely until Swift 6, where this is no longer necessary.

Reply:
> I don't think the user needs to know about the details of the underlying AST.

But the compiler makes decisions based on the AST and Sema decides that (in this situation) that the base is a different declaration, so I'm like to ask again - why would removal of guard statement cause an error in Swift 5 mode?

Reply:
> why would removal of guard statement cause an error in Swift 5 mode?

Because this code is valid and compiles:

```swift
buttonTapped { [weak self] in
  guard let self else { return }
  dismiss() // implicitly refers to `self.dismiss()`
}
```

but this code without the `guard let self` statement is invalid:

```swift
buttonTapped { [weak self] in
  dismiss() // 🛑 call to method 'method' in closure requires explicit use of 'self' to make capture semantics explicit
}
```

Reply:
```
buttonTapped { [weak self] in
  dismiss() // 🛑 call to method 'dismiss' in closure requires explicit use of 'self' to make capture semantics explicit
}
```

Is this well-formed without SE-0365 in Swift 5 mode?

Reply:
No, this doesn't compile in Swift 5.7 if the closure is escaping. It only compiles if the closure is non-escaping (which we're fixing in Swift 6).

Reply:
So in Swift 5 mode without your changes implicit base for `dismiss()` is going to be different depending on whether the closure is escaping or not? Or is it just a matter of how it was checked by MiscDiagnostics?

Reply:
The difference in Swift 5 mode is how it's checked by MiscDiagnostics, yes.

Reply:
Is it possible to encapsulate new checks to Swift 6 mode only?

Reply:
Yes, however this would mean delaying SE-0365 entirely until Swift 6.

I can remove this Swift 5 specific workaround and gate the remaining code for SE-0365 behind Swift 6 if that's best option to unblock things.

Reply:
That sounds reasonable to me!

Reply:
Ok, I reverted the changes related to Swift 5. Now SE-0365 is only enabled in Swift 6 (https://github.com/apple/swift/pull/40702/commits/5bfdfd822c7db7f942a017d5b70308c817c498f2). The code is certainly a lot cleaner now :)